File: day16_part1_2015.py

import re

def parse_input(file_name):
    with open(file_name, 'r') as file:
        data = file.read()
    return data

def check_sue(data):
    mfcsam_data = {
        'children': 3,
        'cats': 7,
        'samoyeds': 2,
        'pomeranians': 3,
        'akitas': 0,
        'vizslas': 0,
        'goldfish': 5,
        'trees': 3,
        'cars': 2,
        'perfumes': 1
    }
    for line in data.split('\n'):
        sue_num = int(re.search(r'\d+', line).group())
        attributes = re.findall(r'(\w+): (\d+)', line)
        sue_data = {attr: int(val) for attr, val in attributes}
        if all(mfcsam_data[attr] == val for attr, val in sue_data.items()):
            return sue_num

data = parse_input("input.txt")
result = check_sue(data)
print(result)

File: objectivec/day16_part1_2015.m
#import <Foundation/Foundation.h>

int main(int argc, const char * argv[]) {
    @autoreleasepool {
        NSString *filePath = @"input.txt";
        NSError *error = nil;
        NSString *fileContents = [NSString stringWithContentsOfFile:filePath encoding:NSUTF8StringEncoding error:&error];
        if (error) {
            NSLog(@"Error reading file: %@", error);
            return 1;
        }
        
        NSArray *lines = [fileContents componentsSeparatedByString:@"\n"];
        NSDictionary *mfcsam = @{
            @"children": @3, @"cats": @7, @"samoyeds": @2, @"pomeranians": @3,
            @"akitas": @0, @"vizslas": @0, @"goldfish": @5, @"trees": @3,
            @"cars": @2, @"perfumes": @1
        };
        
        for (NSString *line in lines) {
            NSArray *parts = [line componentsSeparatedByString:@" "];
            NSString *sueNumber = [[parts[1] substringToIndex:[parts[1] length]-1] mutableCopy];
            
            BOOL matches = YES;
            for (int i = 2; i < [parts count]; i += 2) {
                NSString *item = [[parts[i] substringToIndex:[parts[i] length]-1] mutableCopy];
                NSNumber *count = @([[parts[i+1] substringToIndex:[parts[i+1] length]-1] intValue]);
                
                if (![mfcsam[item] isEqualToNumber:count]) {
                    matches = NO;
                    break;
                }
            }
            
            if (matches) {
                printf("%s\n", [sueNumber UTF8String]);
                break;
            }
        }
    }
    return 0;
}